File: code/機器學習/nn/nn0/nn0c/gpt0.c
#include <stdio.h>
#include <stdlib.h>
#include <math.h>
#include "gpt0.h"

/*
  模型超參數：
    n_layer    — Transformer 層數（深度）
    n_embd     — 嵌入向量維度（寬度）
    block_size — 最大上下文長度（注意力窗口）
    n_head     — 多頭注意力頭數
    head_dim   — 每個注意力頭的維度 = n_embd / n_head
*/
int n_layer = 1;
int n_embd = 16;
int block_size = 16;
int n_head = 4;
int head_dim = 0;

/*
  模型的所有權重矩陣（靜態配置，最多支援 10 層）：
    wte       — token embedding 矩陣 [vocab_size × n_embd]
    wpe       — position embedding 矩陣 [block_size × n_embd]
    lm_head   — 語言模型輸出頭 [vocab_size × n_embd]
    attn_wq/wk/wv/wo — 注意力層的 Q/K/V/O 投影矩陣
    mlp_fc1/fc2      — MLP 的兩層線性變換矩陣
*/
static Matrix wte, wpe, lm_head;
static Matrix attn_wq[10], attn_wk[10], attn_wv[10], attn_wo[10];
static Matrix mlp_fc1[10], mlp_fc2[10];

/*
  初始化 GPT 模型的所有權重。
  使用標準差 0.08 的常態分佈隨機初始化每個矩陣元素。
*/
void init_gpt(int vocab_size) {
    head_dim = n_embd / n_head;
    wte = create_matrix(vocab_size, n_embd, 0.08);
    wpe = create_matrix(block_size, n_embd, 0.08);
    lm_head = create_matrix(vocab_size, n_embd, 0.08);
    for (int i = 0; i < n_layer; i++) {
        attn_wq[i] = create_matrix(n_embd, n_embd, 0.08);
        attn_wk[i] = create_matrix(n_embd, n_embd, 0.08);
        attn_wv[i] = create_matrix(n_embd, n_embd, 0.08);
        attn_wo[i] = create_matrix(n_embd, n_embd, 0.08);
        mlp_fc1[i] = create_matrix(4 * n_embd, n_embd, 0.08);
        mlp_fc2[i] = create_matrix(n_embd, 4 * n_embd, 0.08);
    }
}

/*
  GPT 前向傳播（單一步驟）：
    輸入：當前 token_id、位置 pos_id、KV 快取陣列
    流程：
      1. Token Embedding + Position Embedding
      2. RMS 正規化（Pre-LN 結構）
      3. 對每層 Transformer：
         a. 多頭因果注意力（含 KV Cache 儲存）
         b. 殘差連接
         c. MLP（FC1 → ReLU → FC2）
         d. 殘差連接
      4. lm_head 輸出 logits
*/
Value** gpt_forward(int token_id, int pos_id, Value**** keys, Value**** values) {
    Value** x = arena_alloc(n_embd * sizeof(Value*));
    for (int i = 0; i < n_embd; i++) x[i] = add(wte.data[token_id][i], wpe.data[pos_id][i]);
    x = rmsnorm(x, n_embd);

    for (int li = 0; li < n_layer; li++) {
        // --- 多頭注意力區塊 ---
        Value** x_residual = x;
        x = rmsnorm(x, n_embd);
        Value** q = linear(x, n_embd, attn_wq[li]);
        Value** k = linear(x, n_embd, attn_wk[li]);
        Value** v = linear(x, n_embd, attn_wv[li]);
        
        // 將當前步驟的 K、V 存入快取（供未來位置使用）
        keys[li][pos_id] = k;
        values[li][pos_id] = v;
        
        Value** x_attn = arena_alloc(n_embd * sizeof(Value*));
        int len_seq = pos_id + 1; // 目前為止的序列長度
        
        // 多頭注意力計算
        for (int h = 0; h < n_head; h++) {
            int hs = h * head_dim;
            // Scaled Dot-Product Attention: (Q @ K^T) / sqrt(d_k)
            Value** attn_logits = arena_alloc(len_seq * sizeof(Value*));
            for (int t = 0; t < len_seq; t++) {
                Value* sum = new_value(0.0);
                for (int j = 0; j < head_dim; j++) sum = add(sum, mul(q[hs + j], keys[li][t][hs + j]));
                attn_logits[t] = div_v(sum, new_value(sqrt((double)head_dim)));
            }
            // Softmax 得到注意力權重
            Value** attn_weights = softmax_v(attn_logits, len_seq);
            // 加權求和: attention_weight @ V
            for (int j = 0; j < head_dim; j++) {
                Value* sum = new_value(0.0);
                for (int t = 0; t < len_seq; t++) sum = add(sum, mul(attn_weights[t], values[li][t][hs + j]));
                x_attn[hs + j] = sum;
            }
        }
        // Attention 輸出線性投影 + 殘差連接
        x = linear(x_attn, n_embd, attn_wo[li]);
        Value** x_res1 = arena_alloc(n_embd * sizeof(Value*));
        for (int i = 0; i < n_embd; i++) x_res1[i] = add(x[i], x_residual[i]);
        x = x_res1;
        
        // --- MLP 區塊 ---
        Value** x_residual_mlp = x;
        x = rmsnorm(x, n_embd);
        x = linear(x, n_embd, mlp_fc1[li]);   // 維度擴張：n_embd → 4*n_embd
        Value** x_relu = arena_alloc(4 * n_embd * sizeof(Value*));
        for (int i = 0; i < 4 * n_embd; i++) x_relu[i] = v_relu(x[i]); // ReLU 激活
        x = linear(x_relu, 4 * n_embd, mlp_fc2[li]); // 維度壓縮：4*n_embd → n_embd
        Value** x_res2 = arena_alloc(n_embd * sizeof(Value*));
        for (int i = 0; i < n_embd; i++) x_res2[i] = add(x[i], x_residual_mlp[i]);
        x = x_res2;
    }
    // 輸出層：將最終隱藏狀態映射到詞彙表維度
    return linear(x, n_embd, lm_head);
}

/*
  訓練 GPT 模型：
    流程：
      1. 初始化 Adam 優化器
      2. 對每個訓練步：
         a. 取一個文檔並 tokenize
         b. 初始化 KV 快取
         c. 對序列每個位置做 forward → softmax → cross-entropy loss
         d. 平均所有位置的 loss
         e. 反向傳播計算梯度
         f. Adam 更新參數
         g. 重設記憶體池（清除計算圖暫存節點）
*/
void train_gpt(int num_steps, double learning_rate, char** docs, int num_docs, int vocab_size, int* (*tokenize_cb)(const char*, int*)) {
    init_optimizer();
    
    for (int step = 0; step < num_steps; step++) {
        int out_len;
        int* tokens = tokenize_cb(docs[step % num_docs], &out_len);
        int n = block_size < (out_len - 1) ? block_size : (out_len - 1);

        Value**** keys = arena_alloc(n_layer * sizeof(Value***));
        Value**** values = arena_alloc(n_layer * sizeof(Value***));
        for (int i = 0; i < n_layer; i++) {
            keys[i] = arena_alloc(block_size * sizeof(Value**));
            values[i] = arena_alloc(block_size * sizeof(Value**));
        }

        // 對序列中每個位置計算 loss 並加總
        Value* loss = new_value(0.0);
        for (int pos_id = 0; pos_id < n; pos_id++) {
            Value** logits = gpt_forward(tokens[pos_id], pos_id, keys, values);
            Value** probs = softmax_v(logits, vocab_size);
            // Cross-Entropy Loss = -log(P[target])
            loss = add(loss, neg(v_log(probs[tokens[pos_id + 1]])));
        }
        loss = div_v(loss, new_value((double)n)); // 平均 loss

        // 反向傳播與參數更新
        zero_grad();
        backward(loss);
        step_adam(step, num_steps, learning_rate);

        printf("step %4d / %4d | loss %.4f\r", step + 1, num_steps, loss->data);
        fflush(stdout);
        
        arena_reset(); // 清除計算圖暫存
        free(tokens);
    }
    printf("\n");
}

/*
  推論（文字生成）：
    使用自回歸方式逐 token 生成文字。
    每一步：
      1. 前向傳播得到 logits
      2. Temperature scaling: logits / T
      3. Softmax 轉為機率
      4. 依機率加權取樣下一個 token
      5. 遇到 BOS 或達到最大長度時停止
*/
void inference_gpt(int num_samples, double temperature, int vocab_size, int BOS, char vocab_str[][5]) {
    printf("\n--- inference (new, hallucinated names) ---\n");
    for (int sample_idx = 0; sample_idx < num_samples; sample_idx++) {
        Value**** keys = arena_alloc(n_layer * sizeof(Value***));
        Value**** values = arena_alloc(n_layer * sizeof(Value***));
        for (int i = 0; i < n_layer; i++) {
            keys[i] = arena_alloc(block_size * sizeof(Value**));
            values[i] = arena_alloc(block_size * sizeof(Value**));
        }

        int token_id = BOS;
        printf("sample %2d: ", sample_idx + 1);

        for (int pos_id = 0; pos_id < block_size; pos_id++) {
            Value** logits = gpt_forward(token_id, pos_id, keys, values);
            // Temperature scaling：控制生成的多樣性
            Value** scaled_logits = arena_alloc(vocab_size * sizeof(Value*));
            for (int i = 0; i < vocab_size; i++) {
                scaled_logits[i] = div_v(logits[i], new_value(temperature));
            }
            Value** probs = softmax_v(scaled_logits, vocab_size);
            
            double* probs_data = malloc(vocab_size * sizeof(double));
            for (int i = 0; i < vocab_size; i++) probs_data[i] = probs[i]->data;
            token_id = random_choice(probs_data, vocab_size);
            free(probs_data);

            if (token_id == BOS) break;
            printf("%s", vocab_str[token_id]);
        }
        printf("\n");
        arena_reset();
    }
}

File: code/機器學習/nn/nn0/nn0rs/src/gpt0.rs
//! gpt0.rs — GPT 模型實現（Rust 版）

use std::collections::HashMap;
use std::io::Write;
use rand::prelude::*;
use rand_distr::Normal;

use crate::nn0::{gd, linear, rmsnorm, softmax, Adam, LanguageModel, ValueRef};

/*
  GPT 語言模型結構體。
  儲存模型超參數、所有權重矩陣（state_dict）以及攤平後的參數列表。
*/
pub struct Gpt {
    pub vocab_size: usize,
    pub n_embd: usize,
    pub n_layer: usize,
    pub n_head: usize,
    pub block_size: usize,
    pub head_dim: usize,
    pub state_dict: HashMap<String, Vec<Vec<ValueRef>>>,
    pub params: Vec<ValueRef>,
}

impl Gpt {
    /* 建構函數：初始化所有權重參數 */
    pub fn new(vocab_size: usize, n_embd: usize, n_layer: usize, n_head: usize, block_size: usize) -> Self {
        let head_dim = n_embd / n_head;
        assert_eq!(n_embd % n_head, 0, "n_embd 必須能被 n_head 整除");

        let mut rng = rand::thread_rng();
        let mut state_dict = Self::init_params(vocab_size, n_embd, n_layer, block_size, &mut rng, 0.08);

        // 攤平所有參數供優化器使用
        let mut params = Vec::new();
        for mat in state_dict.values() {
            for row in mat {
                for p in row {
                    params.push(p.clone());
                }
            }
        }

        Self { vocab_size, n_embd, n_layer, n_head, block_size, head_dim, state_dict, params }
    }

    /*
      使用常態分佈 N(0, std) 初始化所有權重矩陣：
        - wte: token embedding
        - wpe: position embedding
        - lm_head: 輸出層
        - 每層的 attn_wq/wk/wv/wo 與 mlp_fc1/fc2
    */
    fn init_params(
        vocab_size: usize, n_embd: usize, n_layer: usize, block_size: usize,
        rng: &mut impl Rng, std: f64,
    ) -> HashMap<String, Vec<Vec<ValueRef>>> {
        let mut sd = HashMap::new();
        let normal = Normal::new(0.0, std).unwrap();

        let mut matrix = |nout: usize, nin: usize| -> Vec<Vec<ValueRef>> {
            (0..nout).map(|_| (0..nin).map(|_| ValueRef::new(normal.sample(rng))).collect()).collect()
        };

        sd.insert("wte".to_string(), matrix(vocab_size, n_embd));
        sd.insert("wpe".to_string(), matrix(block_size, n_embd));
        sd.insert("lm_head".to_string(), matrix(vocab_size, n_embd));

        for i in 0..n_layer {
            sd.insert(format!("layer{}.attn_wq", i), matrix(n_embd, n_embd));
            sd.insert(format!("layer{}.attn_wk", i), matrix(n_embd, n_embd));
            sd.insert(format!("layer{}.attn_wv", i), matrix(n_embd, n_embd));
            sd.insert(format!("layer{}.attn_wo", i), matrix(n_embd, n_embd));
            sd.insert(format!("layer{}.mlp_fc1", i), matrix(4 * n_embd, n_embd));
            sd.insert(format!("layer{}.mlp_fc2", i), matrix(n_embd, 4 * n_embd));
        }
        sd
    }
}

/* 實作 LanguageModel Trait：提供前向傳播 */
impl LanguageModel for Gpt {
    fn block_size(&self) -> usize { self.block_size }
    fn n_layer(&self) -> usize { self.n_layer }

    /*
      前向傳播：
        1. Token Embedding + Position Embedding
        2. RMSNorm → N 層 Transformer（Attention + MLP）
        3. lm_head → logits

      KV Cache 以展平的一維 Vec<ValueRef> 儲存：
        keys[layer] = [K_0, K_1, ..., K_t]（每個 K 長度為 n_embd）
    */
    fn forward(&self, token_id: usize, pos_id: usize, keys: &mut Vec<Vec<ValueRef>>, values: &mut Vec<Vec<ValueRef>>) -> Vec<ValueRef> {
        let sd = &self.state_dict;

        let tok_emb = &sd["wte"][token_id];
        let pos_emb = &sd["wpe"][pos_id];

        let mut x: Vec<ValueRef> = tok_emb.iter().zip(pos_emb.iter()).map(|(t, p)| t.add(p)).collect();
        x = rmsnorm(&x);

        for li in 0..self.n_layer {
            // --- 多頭注意力 ---
            let x_residual = x.clone();
            x = rmsnorm(&x);

            let q = linear(&x, &sd[&format!("layer{}.attn_wq", li)]);
            let k = linear(&x, &sd[&format!("layer{}.attn_wk", li)]);
            let v = linear(&x, &sd[&format!("layer{}.attn_wv", li)]);

            // 將 K/V 展平後追加到該層快取末端
            keys[li].extend(k);
            values[li].extend(v);

            let mut x_attn = Vec::with_capacity(self.n_embd);
            let seq_len = keys[li].len() / self.n_embd;

            for h in 0..self.n_head {
                let hs = h * self.head_dim;
                let q_h = &q[hs..hs + self.head_dim];

                let mut attn_logits = Vec::with_capacity(seq_len);
                for t in 0..seq_len {
                    let k_h_t = &keys[li][t * self.n_embd + hs .. t * self.n_embd + hs + self.head_dim];
                    let dot = q_h.iter().zip(k_h_t.iter())
                        .map(|(q_val, k_val)| q_val.mul(k_val))
                        .reduce(|a, b| a.add(&b)).unwrap();
                    let scale = 1.0 / (self.head_dim as f64).sqrt();
                    attn_logits.push(dot.mul_scalar(scale));
                }

                let attn_weights = softmax(&attn_logits);

                for j in 0..self.head_dim {
                    let sum_v = (0..seq_len)
                        .map(|t| {
                            let v_h_t_j = &values[li][t * self.n_embd + hs + j];
                            attn_weights[t].mul(v_h_t_j)
                        })
                        .reduce(|a, b| a.add(&b)).unwrap();
                    x_attn.push(sum_v);
                }
            }

            let attn_out = linear(&x_attn, &sd[&format!("layer{}.attn_wo", li)]);
            x = attn_out.iter().zip(x_residual.iter()).map(|(a, b)| a.add(b)).collect();

            // --- MLP ---
            let x_residual = x.clone();
            x = rmsnorm(&x);
            x = linear(&x, &sd[&format!("layer{}.mlp_fc1", li)]);
            x = x.into_iter().map(|xi| xi.relu()).collect();
            x = linear(&x, &sd[&format!("layer{}.mlp_fc2", li)]);
            x = x.iter().zip(x_residual.iter()).map(|(a, b)| a.add(b)).collect();
        }

        linear(&x, &sd["lm_head"])
    }
}

/* 訓練函數：遍歷文檔執行梯度下降 */
pub fn train(model: &Gpt, optimizer: &mut Adam, docs: &[&str], uchars: &[char], bos: usize, num_steps: usize) {
    println!("Training for {} steps ...", num_steps);
    for step in 0..num_steps {
        let doc = docs[step % docs.len()];
        let mut tokens = vec![bos];
        for ch in doc.chars() {
            let id = uchars.iter().position(|&c| c == ch).expect("Char not in uchars");
            tokens.push(id);
        }
        tokens.push(bos);

        let loss_val = gd(model, optimizer, &tokens, step, num_steps);
        print!("\rstep {:4} / {:4} | loss {:.4}", step + 1, num_steps, loss_val);
        std::io::stdout().flush().unwrap();
    }
    println!();
}

/*
  推論函數：自回歸文字生成。
  使用 KV Cache 加速（快取每層的 K/V 向量），
  搭配 temperature 控制生成的多樣性。
*/
pub fn inference(model: &Gpt, uchars: &[char], bos: usize, num_samples: usize, temperature: f64) {
    let vocab_size = model.vocab_size;
    println!("--- inference ({} samples, temperature={}) ---", num_samples, temperature);

    let mut rng = rand::thread_rng();

    for sample_idx in 0..num_samples {
        let mut keys: Vec<Vec<ValueRef>> = vec![Vec::new(); model.n_layer()];
        let mut values: Vec<Vec<ValueRef>> = vec![Vec::new(); model.n_layer()];
        let mut token_id = bos;
        let mut sample = String::new();

        for pos_id in 0..model.block_size {
            let logits = model.forward(token_id, pos_id, &mut keys, &mut values);
            let scaled_logits: Vec<ValueRef> = logits.into_iter().map(|l| l.mul_scalar(1.0 / temperature)).collect();
            let probs = softmax(&scaled_logits);
            let p_data: Vec<f64> = probs.iter().map(|p| p.data()).collect();

            let dist = rand::distributions::WeightedIndex::new(&p_data).expect("Failed to create weighted index");
            token_id = dist.sample(&mut rng);

            if token_id == bos { break; }
            sample.push(uchars[token_id]);
        }
        println!("sample {:2}: {}", sample_idx + 1, sample);
    }
}